lib/t/calls.ex

```diff
@@ -292,4 +292,49 @@ defmodule T.Calls do
     |> select([c, p], {c, p})
     |> Repo.all()
   end
+
+  alias T.Media
+
+  def voicemail_s3_url do
+    Media.user_s3_url()
+  end
+
+  def voicemail_url(s3_key) do
+    Media.user_presigned_url(s3_key)
+  end
+
+  def voicemail_upload_form(content_type) do
+    Media.sign_form_upload(
+      key: Ecto.UUID.generate(),
+      content_type: content_type,
+      # 50 MB
+      max_file_size: 50_000_000,
+      expires_in: :timer.hours(1),
+      acl: "private"
+    )
+  end
+
+  @spec voicemail_save_message(uuid, uuid, uuid) ::
+          {:ok, %Voicemail{}} | {:error, reason :: String.t()}
+  def voicemail_save_message(caller_id, match_id, s3_key) do
+    with :ok <- ensure_voicemail_allowed(caller_id, match_id) do
+      Repo.insert!(%Voicemail{caller_id: caller_id, match_id: match_id, s3_key: s3_key})
+    end
+  end
+
+  defp in_match?(user_id, match_id) do
+    Match
+    |> where(id: ^match_id)
+    |> where([m], m.user_id_1 == ^user_id or m.user_id_2 == ^user_id)
+    |> Repo.exists?()
+  end
+
+  @spec ensure_voicemail_allowed(uuid, uuid) :: :ok | {:error, reason :: String.t()}
+  defp ensure_voicemail_allowed(caller_id, match_id) do
+    cond do
+      in_match?(caller_id, match_id) -> :ok
+      # in_live_mode?(caller_id, called_id) -> :ok
```

Comment:
FYI I was also thinking about voicemail being an "invitation" in `live-mode` (but haven't decided yet and probably not in the next release anyways)